File: lib-2/selphi_face_plugin/lib/SelphiFaceConfiguration.dart
import 'SelphiFaceLivenessMode.dart';


class SelphiFaceConfiguration {
  bool mDebug;
  bool mFullscreen;
  bool mCrop;
  double mCropPercent;
  double mSceneTimeout;
  SelphiFaceLivenessMode mLivenessMode;
  double mJPGQuality;
  String mQrValidatorExpression;
  bool mEnableImages;
  String mUTags;
  bool mFrontalCameraPreferred;
  String mLocale;
  bool mStabilizationMode;
  bool mTemplateRawOptimized;
  int mDesiredCameraWidth;
  int mDesiredCameraHeight;
  bool mEnableGenerateTemplateRaw;
  bool mEnableWidgetEventListener;
  bool mQRMode;
  String mTranslationsContent;
  String mViewsContent;
  
  SelphiFaceConfiguration({
    this.mDebug = false,
    this.mFullscreen = true,
    this.mLocale = "",
    this.mCrop = false,
    this.mCropPercent = 1.0,
    this.mSceneTimeout = 0.0,
    this.mEnableImages = false,
    this.mQrValidatorExpression = null,
    this.mLivenessMode = SelphiFaceLivenessMode.LM_PASSIVE,
    this.mUTags = null,
    this.mJPGQuality = 0.95,
    this.mFrontalCameraPreferred = true,
    this.mStabilizationMode = false,
    this.mTemplateRawOptimized = true,
    this.mDesiredCameraWidth = null,
    this.mDesiredCameraHeight = null,
    this.mEnableGenerateTemplateRaw = false,
    this.mEnableWidgetEventListener = false,
    this.mQRMode = false,
    this.mTranslationsContent = "",
    this.mViewsContent = "",
  });
  
  set viewsContent(String viewsContent) {
    mViewsContent = viewsContent;
  }

  String get viewsContent {
    return mViewsContent;
  }
  
  set translationsContent(String translationsContent) {
    mTranslationsContent = translationsContent;
  }

  String get translationsContent {
    return mTranslationsContent;
  }
  
  set qrMode(bool qrMode) {
    mQRMode = qrMode;
  }

  bool get qrMode {
    return mQRMode;
  }
  
  set enableWidgetEventListener(bool enableWidgetEventListener) {
    mEnableWidgetEventListener = enableWidgetEventListener;
  }

  bool get enableWidgetEventListener {
    return mEnableWidgetEventListener;
  }

  set desiredCameraHeight(int desiredCameraHeight) {
    mDesiredCameraHeight = desiredCameraHeight;
  }

  int get desiredCameraHeight {
    return mDesiredCameraHeight;
  }

  set desiredCameraWidth(int desiredCameraWidth) {
    mDesiredCameraWidth = desiredCameraWidth;
  }

  int get desiredCameraWidth {
    return mDesiredCameraWidth;
  }

  set debug(bool debug) {
    mDebug = debug;
  }

  bool get debug {
    return mDebug;
  }

  set fullscreen(bool fullscreen) {
    mFullscreen = fullscreen;
  }

  bool get fullscreen {
    return mFullscreen;
  }


  set locale(String locale) {
    mLocale = locale;
  }

  String get locale {
    return mLocale;
  }

  set crop(bool crop) {
    mCrop = crop;
  }

  bool get crop {
    return mCrop;
  }

  set cropPercent(double cropPercent) {
    mCropPercent = cropPercent;
  }

  double get cropPercent {
    return mCropPercent;
  }

  set sceneTimeout(double sceneTimeout) {
    mSceneTimeout = sceneTimeout;
  }

  double get sceneTimeout {
    return mSceneTimeout;
  }

  set jpgQuality(double jpgQuality) {
    mJPGQuality = jpgQuality;
  }

  double get jpgQuality {
    return mJPGQuality;
  }

  set enableImages(bool enableImages) {
    mEnableImages = enableImages;
  }

  bool get enableImages {
    return mEnableImages;
  }

  set livenessMode(SelphiFaceLivenessMode livenessMode) {
    mLivenessMode = livenessMode;
  }

  SelphiFaceLivenessMode get livenessMode {
    return mLivenessMode;
  }

  set uTags(String uTags) {
    mUTags = uTags;
  }

  String get uTags {
    return mUTags;
  }

  set frontalCameraPreferred(bool frontalCameraPreferred) {
    mFrontalCameraPreferred = frontalCameraPreferred;
  }

  bool get frontalCameraPreferred {
    return mFrontalCameraPreferred;
  }

  set stabilizationMode(bool stabilizationMode) {
    mStabilizationMode = stabilizationMode;
  }

  bool get stabilizationMode {
    return mStabilizationMode;
  }

  set templateRawOptimized(bool templateRawOptimized) {
    mTemplateRawOptimized = templateRawOptimized;
  }

  bool get templateRawOptimized {
    return mTemplateRawOptimized;
  }

  set enableGenerateTemplateRaw(bool enableGenerateTemplateRaw) {
    mEnableGenerateTemplateRaw = enableGenerateTemplateRaw;
  }

  bool get enableGenerateTemplateRaw {
    return mEnableGenerateTemplateRaw;
  }

  factory SelphiFaceConfiguration.fromJson(Map<String, dynamic> parsedJson){
    return SelphiFaceConfiguration(
      mDebug: parsedJson['debug'],
      mCrop: parsedJson ['crop'],
      mCropPercent: parsedJson['cropPercent'],
      mSceneTimeout: parsedJson['sceneTimeout'],
      mJPGQuality: parsedJson['jpgQuality'],
      mEnableImages: parsedJson['enableImages'],
      mLivenessMode: parsedJson ['livenessMode'],
      mQrValidatorExpression: parsedJson['qrValidatorExpression'],
      mFrontalCameraPreferred: parsedJson['frontalCameraPreferred'],
      mStabilizationMode: parsedJson['stabilizationMode'],
      mTemplateRawOptimized: parsedJson['templateRawOptimized'],
      mFullscreen: parsedJson['fullscreen'],
      mUTags: parsedJson['uTags'],
      mLocale: parsedJson['locale'],
      mDesiredCameraHeight: parsedJson['desiredCameraHeight'],
      mDesiredCameraWidth: parsedJson['desiredCameraWidth'],
      mEnableGenerateTemplateRaw: parsedJson['enableGenerateTemplateRaw'],
      mEnableWidgetEventListener: parsedJson['enableWidgetEventListener'],
      mQRMode: parsedJson['qrMode'],
      mTranslationsContent: parsedJson['translationsContent'],
      mViewsContent: parsedJson['viewsContent'],
    );
  }

  Map<String, dynamic> toJson() =>
      {
        'debug': mDebug,
        'crop': mCrop,
        'cropPercent': mCropPercent,
        'sceneTimeout': mSceneTimeout,
        'enableImages': mEnableImages,
        'jpgQuality': mJPGQuality,
        'livenessMode': mLivenessMode.toString(),
        'qrValidatorExpression': mQrValidatorExpression,
        'frontalCameraPreferred': mFrontalCameraPreferred,
        'stabilizationMode': mStabilizationMode,
        'templateRawOptimized': mTemplateRawOptimized,
        'fullscreen': mFullscreen,
        'uTags': mUTags,
        'locale': mLocale,
        'desiredCameraHeight': mDesiredCameraHeight,
        'desiredCameraWidth': mDesiredCameraWidth,
        'enableGenerateTemplateRaw': mEnableGenerateTemplateRaw,
        'enableWidgetEventListener': mEnableWidgetEventListener,
        'qrMode': mQRMode,
        'translationsContent': mTranslationsContent,
        'viewsContent': mViewsContent,
      };
}

File: plugin/selphi_face_plugin/android/src/main/java/com/selphi/flutter/plugin/selphi_face_plugin/SelphiFaceConfiguration.java
package com.selphi.flutter.plugin.selphi_face_plugin;

import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import android.util.Log;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import com.facephi.fphiwidgetcore.WidgetConfiguration;
import com.facephi.fphiwidgetcore.WidgetLivenessMode;
import org.json.JSONException;
import org.json.JSONObject;
import io.flutter.embedding.android.FlutterActivity;
import io.flutter.plugin.common.BasicMessageChannel;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.StringCodec;

public class SelphiFaceConfiguration {
    /**
     * CONSTANTS
     **/
    private static final String SELPHI_LOG_EVENT_NAME = "onSelphiLogEvent";
    private static final String SELPHI_LOG_PARAM_NAME = "selphiLogInfo";
    private static final String SELPHI_LOG_QR_EVENT_NAME = "onSelphiLogQREvent";
    private static final String SELPHI_LOG_QR_PARAM_NAME = "qrData";

    private static final SelphiFaceObservableQR s = new SelphiFaceObservableQR(false);
    private static final SelphiFaceEventQRListener obs = new SelphiFaceEventQRListener();

    protected enum ConfigurationParams {
        CONF_DEBUG("debug"),
        CONF_FULLSCREEN("fullscreen"),
        CONF_FRONTAL_CAMERA_PREFERRED("frontalCameraPreferred"),
        CONF_LOCALE("locale"),
        CONF_LIVENESS_MODE("livenessMode"),
        CONF_ENABLE_IMAGES("enableImages"),
        CONF_SCENE_TIMEOUT("sceneTimeout"),
        CONF_JPG_QUALITY("jpgQuality"),
        CONF_U_TAGS("uTags"),
        CONF_DESIRED_CAMERA_WIDTH("desiredCameraWidth"),
        CONF_DESIRED_CAMERA_HEIGHT("desiredCameraHeight"),
        CONF_STABILIZATION_MODE("stabilizationMode"),
        CONF_TEMPLATE_RAW_OPTIMIZED("templateRawOptimized"),
        CONF_QR_FLAG("qrMode"),
        CONF_CROP_PERCENT("cropPercent"),
        CONF_CROP("crop"),
        CONF_ENABLE_EVENT_LISTENER("enableWidgetEventListener"),
        CONF_TRANSLATIONS_CONTENT("translationsContent"),
        CONF_VIEWS_CONTENT("viewsContent");

        private final String name;

        ConfigurationParams(String name) {
            this.name = name;
        }

        public String getName() {
            return this.name;
        }
    }

    /**
     * Configures the user control operation and launches the activity that will execute it.
     *
     * @return
     */
    public static WidgetConfiguration createWidgetConfiguration(String resourcesPath, JSONObject config, PluginRegistry.Registrar mRegistrar, FlutterActivity mFlutterActivity) throws Exception {
        try {
            WidgetConfiguration conf = new WidgetConfiguration();

            conf.setResourcesPath(resourcesPath + ".zip");

            return processJSONConfiguration(conf, config, mRegistrar, mFlutterActivity);
        } catch (Exception exc) {
            throw exc;
        }
    }

    /**
     * Processes the JSON input argument and sets the configuration of the User Control.
     *
     * @param hybridConfiguration ReadableMap with input JS arguments.
     * @param mRegistrar
     * @param mFlutterActivity
     * @return Configuration of the widget
     */
    private static WidgetConfiguration processJSONConfiguration(WidgetConfiguration widgetConfiguration, JSONObject hybridConfiguration, final PluginRegistry.Registrar mRegistrar, final FlutterActivity mFlutterActivity) throws Exception {
        try {
            JSONObject actualObject = hybridConfiguration;

            if (actualObject.length() == 0) return widgetConfiguration;

            if (actualObject.has(ConfigurationParams.CONF_DEBUG.getName()))
                widgetConfiguration.setDebug(actualObject.getBoolean(ConfigurationParams.CONF_DEBUG.getName()));
            if (actualObject.has(ConfigurationParams.CONF_FULLSCREEN.getName()))
                widgetConfiguration.setFullscreen(actualObject.getBoolean(ConfigurationParams.CONF_FULLSCREEN.getName()));
            if (actualObject.has(ConfigurationParams.CONF_LOCALE.getName()))
                widgetConfiguration.setLocale(actualObject.getString(ConfigurationParams.CONF_LOCALE.getName()));
            if (actualObject.has(ConfigurationParams.CONF_ENABLE_IMAGES.getName()))
                widgetConfiguration.logImages(actualObject.optBoolean(ConfigurationParams.CONF_ENABLE_IMAGES.getName(), true));
            if (actualObject.has(ConfigurationParams.CONF_SCENE_TIMEOUT.getName()))
                widgetConfiguration.setSceneTimeout((float) actualObject.optDouble(ConfigurationParams.CONF_SCENE_TIMEOUT.getName(), 0.0));
            if (actualObject.has(ConfigurationParams.CONF_JPG_QUALITY.getName()))
                widgetConfiguration.setJPGQuality((float) actualObject.optDouble(ConfigurationParams.CONF_JPG_QUALITY.getName(), 0.8f));
            if (actualObject.has(ConfigurationParams.CONF_DESIRED_CAMERA_WIDTH.getName()))
                widgetConfiguration.setParam("DesiredCameraWidth", String.valueOf(actualObject.optInt(ConfigurationParams.CONF_DESIRED_CAMERA_WIDTH.getName())));
            if (actualObject.has(ConfigurationParams.CONF_DESIRED_CAMERA_HEIGHT.getName()))
                widgetConfiguration.setParam("DesiredCameraHeight", String.valueOf(actualObject.optInt(ConfigurationParams.CONF_DESIRED_CAMERA_HEIGHT.getName())));
            if (actualObject.has(ConfigurationParams.CONF_STABILIZATION_MODE.getName()))
                widgetConfiguration.setStabilizationMode(actualObject.optBoolean(ConfigurationParams.CONF_STABILIZATION_MODE.getName(), false));
            if (actualObject.has(ConfigurationParams.CONF_TEMPLATE_RAW_OPTIMIZED.getName()))
                widgetConfiguration.setTemplateRawOptimized(actualObject.optBoolean(ConfigurationParams.CONF_TEMPLATE_RAW_OPTIMIZED.getName(), false));
            if (actualObject.has(ConfigurationParams.CONF_QR_FLAG.getName()) &&
                    actualObject.getBoolean(ConfigurationParams.CONF_QR_FLAG.getName())) {
                widgetConfiguration.setQRFlag(actualObject.optBoolean(ConfigurationParams.CONF_QR_FLAG.getName(), false));
                StartEventQRListener(widgetConfiguration, mRegistrar, mFlutterActivity);
            }
            if (actualObject.has(ConfigurationParams.CONF_FRONTAL_CAMERA_PREFERRED.getName())) {
                boolean isFrontalCamera = actualObject.optBoolean(ConfigurationParams.CONF_FRONTAL_CAMERA_PREFERRED.getName(), true);
                if (isFrontalCamera) widgetConfiguration.setFrontFacingCameraAsPreferred();
                else widgetConfiguration.setBackFacingCameraAsPreferred();
            }

            if (actualObject.has(ConfigurationParams.CONF_LIVENESS_MODE.getName())) {
                String livenessMode = actualObject.optString(ConfigurationParams.CONF_LIVENESS_MODE.getName());
                if (livenessMode.equalsIgnoreCase("PASSIVE")) {
                    widgetConfiguration.setParam("DesiredCameraWidth", "1280");
                    widgetConfiguration.setParam("DesiredCameraHeight", "720");
                    widgetConfiguration.setLivenessMode(WidgetLivenessMode.LIVENESS_PASSIVE);
                } else widgetConfiguration.setLivenessMode(WidgetLivenessMode.LIVENESS_NONE);
            }

            if (actualObject.has(ConfigurationParams.CONF_U_TAGS.getName())) {
                String userTagsStr = actualObject.optString("uTags", null);
                byte[] userTags;
                if (userTagsStr != null && !userTagsStr.isEmpty()) {
                    userTags = Base64.decode(userTagsStr, Base64.DEFAULT);
                    widgetConfiguration.setUserTags(userTags);
                }
            }

            if (actualObject.has(ConfigurationParams.CONF_ENABLE_EVENT_LISTENER.getName()) &&
                    actualObject.getBoolean(ConfigurationParams.CONF_ENABLE_EVENT_LISTENER.getName())) {
                StartEventListener(widgetConfiguration, mRegistrar, (FlutterActivity) mFlutterActivity);
            }

            if (actualObject.has(ConfigurationParams.CONF_VIEWS_CONTENT.getName())) {
                String viewsContent = actualObject.optString(ConfigurationParams.CONF_VIEWS_CONTENT.getName(), "");
                if (!viewsContent.equals(""))
                    widgetConfiguration.setViewsContent(viewsContent);
            }

            if (actualObject.has(ConfigurationParams.CONF_TRANSLATIONS_CONTENT.getName())) {
                String translationsContent = actualObject.optString(ConfigurationParams.CONF_TRANSLATIONS_CONTENT.getName(), "");
                if (!translationsContent.equals(""))
                    widgetConfiguration.setTranslationsContent(translationsContent);
            }

            if (actualObject.has(ConfigurationParams.CONF_CROP_PERCENT.getName()))
                if (widgetConfiguration.getExtractionConfig() != null)
                    widgetConfiguration.getExtractionConfig().setCropImagePercent((float) actualObject.getDouble(ConfigurationParams.CONF_CROP_PERCENT.getName()));
            if (actualObject.has(ConfigurationParams.CONF_CROP.getName()))
                if (widgetConfiguration.getExtractionConfig() != null)
                    widgetConfiguration.getExtractionConfig().setCropImageDebug(actualObject.getBoolean(ConfigurationParams.CONF_CROP.getName()));
        } catch (Exception exc) {
            throw exc;
        }
        return widgetConfiguration;
    }

    private static void StartEventQRListener(WidgetConfiguration widgetConfiguration, final PluginRegistry.Registrar mRegistrar, final FlutterActivity mFlutterActivity) {
        widgetConfiguration.setIFPhiWidgetQR_classname("com.selphi.flutter.plugin.selphi_face_plugin.SelphiFaceEventQRListener");

        s.addObserver(obs); // Add the Observer
        if (s.getIsValid()) {
            // IMPORTANT. // Make changes to the Subject.
            s.setIsValid(false);
        }

        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                SelphiFaceLogQRModel.getLogModel().getCurrentLog().observe(mFlutterActivity, new Observer<String>() {
                    @Override
                    public void onChanged(String logData) {
                        JSONObject params = new JSONObject();
                        try {
                            params.put(SELPHI_LOG_QR_PARAM_NAME, logData);
                        } catch (JSONException e) {
                            Log.d("Event Exception:", e.getMessage());
                        }

                        BasicMessageChannel channel = new BasicMessageChannel(mRegistrar.messenger(), SELPHI_LOG_QR_EVENT_NAME, StringCodec.INSTANCE);
                        channel.send(params.toString(), new BasicMessageChannel.Reply() {
                            @Override
                            public void reply(@Nullable Object reply) {
                                Log.i("obj", reply.toString());
                                if (reply.equals("true")) {
                                    // Make changes to the Subject.
                                    s.setIsValid(true);
                                }
                            }
                        });
                    }
                });
            }
        });
    }

    private static void StartEventListener(WidgetConfiguration widgetConfiguration, final PluginRegistry.Registrar mRegistrar, final FlutterActivity mFlutterActivity) {
        widgetConfiguration.setIFPhiWidgetEventListener_classname("com.selphi.flutter.plugin.selphi_face_plugin.SelphiFaceEventListener");
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                SelphiFaceLogModel.getLogModel().getCurrentLog().observe(mFlutterActivity, new Observer<String>() {
                    @Override
                    public void onChanged(String logData) {
                        JSONObject params = new JSONObject();
                        try {
                            params.put(SELPHI_LOG_PARAM_NAME, logData);
                        } catch (JSONException e) {
                            Log.d("Event Exception:", e.getMessage());
                        }

                        BasicMessageChannel channel = new BasicMessageChannel(mRegistrar.messenger(), SELPHI_LOG_EVENT_NAME, StringCodec.INSTANCE);
                        channel.send(params.toString());
                    }
                });
            }
        });
    }
}